### DISCUSSION TITLE
Fitur Baru: Interactive Message - Rilis v1.0-RC21

### DISCUSSION THREAD

**Discussion by rioastamal** [kirimwa-id/forum#167] [Pengumuman]
Halo semua,

Dengan senang hati saya mengumumkan jika fitur _interactive message_ #143 sekarang sudah bisa digunakan. Ini juga menandai rilis v1.0-RC21. 

Interactive message adalah pesan yang dapat memberikan panduan pilihan (opsi) kepada penerima sehingga format balasan sesuai yang diinginkan. Terdapat dua tipe interactive message:

- List Message
- Reply Button

Dokumentasi API interactive message ada di https://developer.kirimwa.id/api/v1/#kirim-pesan-interaktif.

> **PENTING**  \
Fitur ini dapat digunakan oleh semua member API KirimWA.id hingga **10 April 2023**. Setelah tanggal tersebut fitur ini hanya bisa digunakan member Pro dan Enterprise saja.

### List Message

Gunakan list message jika ingin memberikan maksimal hingga 10 opsi kepada penerima. Pada atribut `message_type` gunakan nilai `list` untuk mengirim list message.

```sh
LIST_MESSAGE="$(cat <<EOF
{
  "phone_number": "6281234567890-1628639662",
  "message_type": "list",
  "device_id": "iphone-x-pro",
  "is_group_message": true,
  "message": {
    "title": "Menu Depot ABC",
    "description": "Sedia berbagai makanan dan minuman khas Indonesia",
    "footer": "visit https://developer.kirimwa.id",
    "label_menu": "Menu",
    "list": [
      {
        "title": "Makanan",
        "items": [
          {
            "id": "soto_ayam",
            "label": "Soto Ayam Lamongan",
            "description": "Perpaduan ayam kampung dan bumbu yang khas"
          },
          {
            "id": "rawon",
            "label": "Rawon",
            "description": "Sup dading berwarna hitam dengan bumbu kluwek"
          },
          {
            "id": "sate_kambing",
            "label": "Sate Kambing",
            "description": "Daging kambing yang empuk dan gurih dengan bumbu kecap atau kacang"
          }
        ]
      },
      {
        "title": "Minuman",
        "items": [
          {
            "id": "es_teh",
            "label": "Es Teh",
            "description": "Es teh dari kebun teh pilihan"
          },
          {
            "id": "jus_jeruk",
            "label": "Jus Jeruk",
            "description": "Jus jeruk segar dan dingin"
          },
          {
            "id": "air_putih",
            "label": "Air Putih"
          }
        ]
      }
    ]
  }
}
EOF
)"
```

```sh
curl -X POST 'https://api.kirimwa.id/v1/messages' \
  -H 'Authorization: Bearer API_TOKEN' \
  -H 'Content-Type: application/json' \
  -d "$LIST_MESSAGE"
```

![kirimwa-rc21-00](https://user-images.githubusercontent.com/469847/220904536-24fa12ab-d30d-481a-a8e7-5c000706ecad.jpg)
![kirimwa-rc21-01](https://user-images.githubusercontent.com/469847/220904569-d4aac8bf-dd9f-492c-b5b6-a2e8e821e2d2.jpg)

### Reply Button

Gunakan reply button jika ingin memberikan maksimal hingga 3 opsi kepada penerima. Pada atribut `message_type` gunakan nilai `buttons` untuk mengirim.

```sh
BUTTONS_MESSAGE="$(cat <<EOF
{
  "phone_number": "6281234567890-1628639662",
  "message_type": "buttons",
  "device_id": "iphone-x-pro",
  "is_group_message": true,
  "message": {
    "description": "Apa warna favoritmu?",
    "footer": "visit https://developer.kirimwa.id",
    "buttons": [
      {
        "id": "merah",
        "label": "Merah"
      },
      {
        "id": "hijau",
        "label": "Hijau"
      },
      {
        "id": "biru",
        "label": "Biru"
      }
    ]
  }
}
EOF
)"
```

```sh
curl -X POST 'https://api.kirimwa.id/v1/messages' \
  -H 'Authorization: Bearer API_TOKEN' \
  -H 'Content-Type: application/json' \
  -d "$BUTTONS_MESSAGE"
```

![kirimwa-rc21-02](https://user-images.githubusercontent.com/469847/220913371-68516531-90c8-4f7a-95af-e7074b36562c.jpg)

### Memproses Respon

Untuk memproses respon pastikan anda mengaktifkan [webhooks](https://developer.kirimwa.id/api/v1/#webhooks). Gunakan atribut pada `payload.message_type` untuk menentukan tipe dari sebuah pesan yang masuk. Untuk list message nilainya adalah `list_response` dan untuk reply button nilainya adalah `buttons_response`.

Sedangkan untuk mengakses nilai yang dipilih gunakan atribut `payload.interactive_message`.

Berikut ini contoh respon dari pilihan sebuah list message dari respon diatas yaitu penerima memilih "Rawon".

```json
{
  "id": "A78E81E77A91A1A5808CC8A44FACCA12",
  "webhook_type": "incoming_message",
  "payload": {
    "id": "A78E81E77A91A1A5808CC8A44FACCA12",
    "device_id": "iphone-x-pro",
    "sender": "6281234567890",
    "from_me": false,
    "text": null,
    "caption": null,
    "message_type": "list_response",
    "is_group_message": true,
    "group_id": "6281234567890-1628639662",
    "timestamp": 1677152770,
    "contact": {
      "name": null,
      "vcard": null
    },
    "location": {
      "latitude": 0,
      "longitude": 0,
      "name": null,
      "address": null
    },
    "meta": {
      "file_name": null,
      "seconds": 0,
      "size": 0,
      "mime_type": "application/json",
      "dimension": {
        "height": 0,
        "width": 0
      }
    },
    "interactive_response": {
      "id": "rawon",
      "label": "Rawon"
    }
  },
  "created_at": "2023-02-23T11:46:10.000Z",
  "server_time": "2023-02-23T11:46:10.713Z"
}
```

Berikut ini contoh respon dari pilihan sebuah reply button dari respon diatas yaitu penerima memilih "Hijau".

```json
{
  "id": "F914269DB171CB4A9E1CD9CBDF00039E",
  "webhook_type": "incoming_message",
  "payload": {
    "id": "F914269DB171CB4A9E1CD9CBDF00039E",
    "device_id": "iphone-x-pro",
    "sender": "6281234567890",
    "from_me": false,
    "text": null,
    "caption": null,
    "message_type": "buttons_response",
    "is_group_message": true,
    "group_id": "6281234567890-1628639662",
    "timestamp": 1677153598,
    "contact": {
      "name": null,
      "vcard": null
    },
    "location": {
      "latitude": 0,
      "longitude": 0,
      "name": null,
      "address": null
    },
    "meta": {
      "file_name": null,
      "seconds": 0,
      "size": 0,
      "mime_type": "application/json",
      "dimension": {
        "height": 0,
        "width": 0
      }
    },
    "interactive_response": {
      "id": "hijau",
      "label": "Hijau"
    }
  },
  "created_at": "2023-02-23T11:59:58.000Z",
  "server_time": "2023-02-23T11:59:59.154Z"
}
```

Saya harap dengan fitur ini akan mempermudah teman-teman developer dalam mengembangkan aplikasi yang mengandalkan API KirimWA.id.

Belum memiliki Token API KirimWA.id? Daftar gratis di https://developer.kirimwa.id/

Salam,
Rio Astamal

**Comment by sifuim** [comment]
Mantap!!!

**Comment by fahram** [comment]
Semoga bisa ditambahkan Button tipe Link

**Reply by rioastamal** [reply]
Akan dimasukkan roadmap

**Comment by sanusi86** [comment]
Maaf mas Rio,
Bisa untuk Url Button dan Call Button gak

**Reply by rioastamal** [reply]
Untuk saat ini belum, akan dimasukkan roadmap

**Comment by rickyok** [comment]
Sepertinya hari ini saya coba, kedua list dan button tidak bekerja di android dan ios

hanya bisa di muncul di wa web saja. Kalau di android dan ios message nya tidak sampai jadinya

**Reply by irfanandratama** [reply]
> Sepertinya hari ini saya coba, kedua list dan button tidak bekerja di android dan ios
> 
> hanya bisa di muncul di wa web saja. Kalau di android dan ios message nya tidak sampai jadinya

Sepertinya memang dari suatu update di aplikasi Whatsapp di perangkat mobile yang membuat fitur tersebut nggak bisa digunakan 
https://github.com/pedroslopez/whatsapp-web.js/issues/1759#issue-1424177626

**Reply by pratamaahady** [reply]
ada update kah mas terkait issue ini ?

**Reply by rioastamal** [reply]
Sepertinya ada semacam pengecekan disisi WhatsApp server, jika memang butuh Buttons dan List sebaiknya menggunakan Official API
